File: JavaScript/boj/Gold/9663.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? 'dev/stdin' : './input.txt';
const N = Number(fs.readFileSync(filePath).toString().trim());

const queen = [];
let result = 0;

const isPossible = (row, col) => {
  for (let i = 0; i < queen.length; i++) {
    const [x, y] = queen[i];

    if (row === x || col === y || Math.abs(row - x) === Math.abs(col - y)) {
      return false;
    }
  }

  return true;
};

const dfs = (row) => {
  if (row === N) {
    result++;
    return;
  }

  for (let col = 0; col < N; col++) {
    if (!isPossible(row, col)) continue;

    queen.push([row, col]);
    dfs(row + 1);
    queen.pop();
  }
};

dfs(0);
console.log(result);

File: Python/boj/Gold/9663.py
import sys

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N = int(input())
    result = 0
    cols = [False] * N
    diagonal_left = [False] * (2 * N - 1)
    diagonal_right = [False] * (2 * N - 1)

    def dfs(row):
        global result

        if row == N:
            result += 1
            return

        for col in range(N):
            if (
                cols[col]
                or diagonal_left[row + col]
                or diagonal_right[row - col + N - 1]
            ):
                continue

            cols[col] = True
            diagonal_left[row + col] = True
            diagonal_right[row - col + N - 1] = True

            dfs(row + 1)

            cols[col] = False
            diagonal_left[row + col] = False
            diagonal_right[row - col + N - 1] = False

    dfs(0)
    print(result)
